cranelift/codegen/src/peepmatic.rs

```diff
@@ -681,39 +651,290 @@ unsafe impl<'a, 'b> InstructionSet<'b> for &'a dyn TargetIsa {
+                opcode: opcode @ Opcode::Trapz,
+                arg,
+                code: _,
+            } => {
+                operands.extend(iter::once(Part::Instruction(arg.into())));
+                opcode
+            }
+
+            InstructionData::IntCompare {
+                opcode: opcode @ Opcode::Icmp,
+                cond,
+                args,
+            } => {
+                operands.extend(
+                    iter::once(intcc_to_peepmatic(cond).into())
+                        .chain(args.iter().map(|v| Part::Instruction((*v).into()))),
+                );
+                opcode
+            }
+
+            InstructionData::IntCompareImm {
+                opcode: opcode @ Opcode::IcmpImm,
+                cond,
+                imm,
+                arg,
+            } => {
+                operands.extend(
+                    iter::once(intcc_to_peepmatic(cond).into())
+                        .chain(iter::once(Part::Constant(imm.into())))
+                        .chain(iter::once(Part::Instruction(arg.into()))),
+                );
+                opcode
+            }
+
+            InstructionData::Ternary {
+                opcode: opcode @ Opcode::Select,
+                ref args,
+            } => {
+                operands.extend(args.iter().map(|v| Part::Instruction((*v).into())));
+                opcode
+            }
+
+            InstructionData::Unary {
+                opcode: opcode @ Opcode::AdjustSpDown,
+                arg,
+            }
+            | InstructionData::Unary {
+                opcode: opcode @ Opcode::Bint,
+                arg,
+            }
+            | InstructionData::Unary {
+                opcode: opcode @ Opcode::Ireduce,
+                arg,
+            }
+            | InstructionData::Unary {
+                opcode: opcode @ Opcode::Sextend,
+                arg,
+            }
+            | InstructionData::Unary {
+                opcode: opcode @ Opcode::Uextend,
+                arg,
+            } => {
+                operands.extend(iter::once(Part::Instruction(arg.into())));
+                opcode
+            }
+
+            InstructionData::UnaryBool { opcode, imm } => {
+                operands.extend(iter::once(Part::Constant(Constant::Bool(
+                    imm,
+                    BitWidth::Polymorphic,
+                ))));
+                opcode
+            }
+
+            InstructionData::UnaryImm {
+                opcode: opcode @ Opcode::AdjustSpDownImm,
+                imm,
+            }
+            | InstructionData::UnaryImm {
+                opcode: opcode @ Opcode::Iconst,
+                imm,
+            } => {
+                operands.extend(iter::once(imm.into()));
+                opcode
+            }
+            ref otherwise => {
+                log::trace!("Not supported by Peepmatic: {:?}", otherwise);
+                return None;
+            }
+        })
```

Comment:
I mean, I think I understand why this is necessary but it is a shame not to be able to use [opcode](https://docs.rs/cranelift-codegen/0.66.0/cranelift_codegen/ir/instructions/enum.InstructionData.html#method.opcode) and [arguments](https://docs.rs/cranelift-codegen/0.66.0/cranelift_codegen/ir/instructions/enum.InstructionData.html#method.arguments) off of `InstructionData` to make things a bit less verbose.

Reply:
There are two reasons it is the way it currently is:

1. As written, we avoid matching on the instruction data twice. I've found that sometimes LLVM can collapse these back-to-back matches into a single match, but not always and it's kind of fragile.

2. The branching instructions supported by peepmatic ignore the block arguments, so we can't use `arguments` directly, we would still need to special case branches.